text/0814-stylelint.md

```diff
@@ -0,0 +1,180 @@
+
+[Stylelint](https://github.com/stylelint/stylelint) is a modern linter for stylesheets that ships with 
+over 170 built-in rules, sane default configs, and robust plugin support.
+
+## Detailed design
+
+The general idea is that we will update the `app` and `addon` [blueprints](https://github.com/ember-cli/ember-cli/tree/master/blueprints) 
+to add a few Stylelint-related packages to the `package.json`, add config files to enable this tooling 
+by default, and add commands to the `scripts` section of `package.json` to expose easy access to this 
+tool consistent with our existing linters.
+
+### Packages
+
+The following packages will be added to the `devDependencies` section of both blueprints:
+
+* [stylelint](https://github.com/stylelint/stylelint) - The base linter
+* [stylelint-config-standard](https://github.com/stylelint/stylelint-config-standard) - The standard rule config supported by the stylelint community
+* [stylelint-prettier](https://github.com/prettier/stylelint-prettier) - Plugin to support prettier auto-formatting as stylelint rules
+* [stylelint-config-prettier](https://github.com/prettier/stylelint-config-prettier) - Plugin to disable stylelint rules that conflict with prettier's automatic behavior
+
+Prettier itself is already included in the Ember blueprints by default.
+
+### Configuration Changes
+
+A `.stylelintrc.js` file will be added with the following contents:
+
+```js title=".stylelintrc.js"
+'use strict';
+
+module.exports = {
+  extends: ['stylelint-config-standard', 'stylelint-prettier/recommended'],
+};
+```
+
+There are many rules, configs, and plugins available for stylelint and prettier, but we are 
+intentionally avoiding being prescriptive on anything that isn't nearly universally agreed upon 
+by the stylelint, prettier, and Ember communities.
+
+#### Lint Ignores
+
+A `.stylelintignore` file will be added with the following contents:
+
+```title=".stylelintignore"
+# unconventional js
+/blueprints/*/files/
+/vendor/
+
+# compiled output
+/dist/
+/tmp/
+
+# dependencies
+/bower_components/
+/node_modules/
+
+# misc
+/coverage/
+!.*
+.*/
+.eslintcache
+.lint-todo/
+
+# ember-try
+/.node_modules.ember-try/
+/bower.json.ember-try
+/npm-shrinkwrap.json.ember-try
+/package.json.ember-try
+/package-lock.json.ember-try
+/yarn.lock.ember-try
+```
+
+These contents are based on the existing `.eslintignore` and `.prettierignore` files for consistency.
+
+### Blueprint Changes
+
+In general, it is recommended that all blueprints provided by addons should satisfy the default 
+linting configuration of a new Ember application. As such, the blueprints provided by ember-cli 
+will be updated as needed to ensure that they satisfy these new linting requirements.
+
+#### `package.json` scripts
+
+The `app` and `addon` blueprints will be updated to add style-related linting scripts:
+
+```json title="package.json"
+{
+  "scripts": {
+    "lint:css": "stylelint \"**/*.{css}\"",
```

Comment:
Should the braces include the possibly CSS extensions?
```suggestion
    "lint:css": "stylelint \"**/*.{css,sass,scss}\"",
```

Reply:
The Stylelint config suggested in this RFC only supports plain `css` files.
If you do try to lint e.g. a `sass` file with this config, you would get:

```
When linting something other than CSS, you should install an appropriate syntax, e.g. "postcss-scss", and use the "customSyntax" option
```

Therefore, it seemed more logical to me to only include `css`.

Reply:
2 cents from my side on this:
out of the box ESLint ember config (as well as Prettier) includes only `*.js` files. If addon or app author decides to add TypeScript support, ESLint config needs to be edited to run ESLint for `*.ts` files.

Hence leaving only `*.css` would be in line with what we do for JavaScript by requiring app/addon author to add sass/scss/etc. file extension should it be added to the codebase.

Reply:
If we only support pure CSS out-of-the-box, then I suppose covering only CSS here is fine.

However, it would definitely be nicer if we could avoid mention file extensions at all somehow, the same way JS linting just uses `eslint .`.

Reply:
Would be nicer indeed, but don't think we can at the moment. Stylelint seems to include everything it can find when running `npx stylelint .`.

Reply:
Yes, in my testing calling `stylelint .` produces very weird results, so we need to scope it to the file extension.